File: csharp/src/Ice/ValueFactoryManager.cs
// **********************************************************************
//
// Copyright (c) 2003-present ZeroC, Inc. All rights reserved.
//
// **********************************************************************

namespace IceInternal
{

    using System.Collections.Generic;

    public sealed class ValueFactoryManagerI : Ice.ValueFactoryManager
    {
        public void add(Ice.ValueFactory factory, string id)
        {
            lock(this)
            {
                if(_factoryMap.ContainsKey(id))
                {
                    Ice.AlreadyRegisteredException ex = new Ice.AlreadyRegisteredException();
                    ex.id = id;
                    ex.kindOfObject = "value factory";
                    throw ex;
                }
                _factoryMap[id] = factory;
            }
        }

        public Ice.ValueFactory find(string id)
        {
            lock(this)
            {
                Ice.ValueFactory factory = null;
                _factoryMap.TryGetValue(id, out factory);
                return factory;
            }
        }

        private Dictionary<string, Ice.ValueFactory> _factoryMap = new Dictionary<string, Ice.ValueFactory>();
    }

}

File: js/src/Ice/ValueFactoryManager.js
// Copyright (c) ZeroC, Inc.

import { AlreadyRegisteredException } from "./LocalExceptions.js";

// Only for use by Instance
export class ValueFactoryManager {
    constructor() {
        this._factoryMap = new Map(); // Map<String, ValueFactory>
    }

    add(factory, id) {
        if (this._factoryMap.has(id)) {
            throw new AlreadyRegisteredException("value factory", id);
        }
        this._factoryMap.set(id, factory);
    }

    find(id) {
        return this._factoryMap.get(id);
    }

    destroy() {
        this._factoryMap = new Map(); // Map<String, ValueFactory>
    }
}
